## ひろしまQuest2020#stayhome【球種予測部門】
#### all_merge_18
- コースの種類、1球前のボール、ファウル数

#### all_merge_17
- イニング時間、打席時間、平均投球間隔(差)、サヨナラ、延長戦

#### all_merge_16
- 球団ごとの年棒ランク
- 選手IDごとの2017年のコース実績

#### all_merge_15
- 左右で球種/コースを別に集計

#### all_merge_14
- 左右で球種を別に集計

#### all_merge_13
- 前の投球からの時間差

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_18.f'
SUBMIT = 'submit/ball_18.csv'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 217)

In [4]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 217)


No  ball  course  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0                  1       1                  1   
1   1   0.0     8.0                  2       1                  1   
2   2   0.0     5.0                  3       1                  1   
3   3   0.0    12.0                  4       1                  2   
4   4   0.0     8.0                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_straight  bc_curve  \
0        0         0           0       0.0     1.0     0.464734  0.110359   
1        0         0           1       0.0     1.0     0.442351  0.093302   
2        0         0           2       0.0     1.0     0.504212  0.069963   
3        1         0           0       0.0     0.0     0.475961  0.101792   
4        1         0           1       0.0     0.0     0.463496  0.099763   

   bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  \
0   0.151086  0.097383  0.067492     0.043210   0.012933    0.052803   
1   0.109348  0.078518  0.141711     0.065447   0.017669    0.051654   
2   0.065568  0.046520  0.211722     0.059158   0.019231    0.023626   
3   0.226048  0.066617  0.042131     0.018345   0.003343    0.065762   
4   0.194714  0.067542  0.091198     0.023901   0.004486    0.054898   

   bc_course00  bc_course01  bc_course02  bc_course03  bc_course04  \
0     0.064837     0.095199     0.082009     0.040512     0.063252   
1     0.044893     0.071306     0.068422     0.032002     0.043180   
2     0.027656     0.045421     0.045238     0.025275     0.023810   
3     0.034319     0.043647     0.027090     0.054336     0.065374   
4     0.024390     0.034832     0.025124     0.039481     0.047557   

   bc_course05  bc_course06  bc_course07  bc_course08  bc_course09  \
0     0.053916     0.030534     0.046508     0.038328     0.163762   
1     0.036149     0.029388     0.042910     0.033625     0.152078   
2     0.027106     0.023077     0.032784     0.022711     0.202930   
3     0.045046     0.062653     0.097944     0.085507     0.098410   
4     0.033526     0.049596     0.080512     0.069337     0.085243   

   bc_course10  bc_course11  bc_course12  bc_high_str  bc_high_ball  \
0     0.057728     0.130872     0.132543     0.135883      0.221489   
1     0.078157     0.203642     0.164248     0.106283      0.230235   
2     0.081319     0.285165     0.157509     0.076007      0.284249   
3     0.098527     0.055307     0.231840     0.151308      0.196937   
4     0.121054     0.076923     0.312424     0.113468      0.206297   

   bc_mid_str  bc_low_str  bc_low_ball  bc_left_str  bc_left_ball  \
0    0.204959    0.174254     0.263415     0.242045      0.294634   
1    0.157397    0.138195     0.367890     0.184621      0.355720   
2    0.102015    0.095055     0.442674     0.118315      0.488095   
3    0.206965    0.157643     0.287147     0.105057      0.153718   
4    0.162901    0.127988     0.389347     0.084346      0.162167   

   bc_center_str  bc_right_str  ...  salary_year_bat  salary_x_y

### train
- 行数: 257117

In [5]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 217)

### test
- 行数: 521650

In [6]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 217)

In [7]:
del all_pitch
gc.collect()

60

In [8]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [9]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

## Dataset作成

In [10]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [11]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 1.14045 + 0.00244957
[200]	cv_agg's multi_logloss: 1.11886 + 0.00260181
[300]	cv_agg's multi_logloss: 1.11239 + 0.0026026
[400]	cv_agg's multi_logloss: 1.10851 + 0.00260437
[500]	cv_agg's multi_logloss: 1.10582 + 0.00267718
[600]	cv_agg's multi_logloss: 1.10392 + 0.00275679
[700]	cv_agg's multi_logloss: 1.10249 + 0.00281053
[800]	cv_agg's multi_logloss: 1.10159 + 0.00285483
[900]	cv_agg's multi_logloss: 1.10091 + 0.00297007
[1000]	cv_agg's multi_logloss: 1.10061 + 0.0029753
[1100]	cv_agg's multi_logloss: 1.10043 + 0.00305723
[1200]	cv_agg's multi_logloss: 1.10049 + 0.00309928
Best num_boost_round: 1102
Best CV score: 1.1004234715781385
CPU times: user 1h 48min 57s, sys: 19.9 s, total: 1h 49min 17s
Wall time: 13min 42s


In [12]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [13]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 34min 9s, sys: 5.29 s, total: 34min 14s
Wall time: 4min 16s


### Feature Importance

In [14]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
151           height_bat      1362
189  salary_year_dif_p-b      1398
201          bmi_dif_p-c      1401
119                 fork      1401
116                curve      1406
5     player_cnt_in_game      1416
168         bat_game_cnt      1551
74            point_diff      1560
166              bmi_bat      1578
192  rank_x_year_dif_p-b      1583
205  salary_year_dif_b-c      1637
207    rank_year_dif_b-c      1649
208  rank_x_year_dif_b-c      1748
117               slider      1842
167           batter_cnt      1891
193          bmi_dif_p-b      2035
209          bmi_dif_b-c      2101
0      pitch_cnt_in_game      2125
56           elapsed_min      2183
59   elapsed_from_inning      2412
61        elapsed_batter      2424
6    pitcher_cnt_in_game      2435
64    ave_elapsed_batter      2661
57              min_diff      2766
65   diff_elapsed_batter      2807
62      ave_elapsed_game      3074
63    ave_elapsed_inning      3086
67     diff_elapsed_game      3152
66   diff_elapsed_inning      3182
55     date_from_opening      3897

In [15]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 6min 17s, sys: 844 ms, total: 6min 18s
Wall time: 48.8 s


## 結果出力

In [16]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.497247  0.044221  0.151861  0.207383  0.005027  0.000025   
1      1  0.449866  0.151382  0.168919  0.172314  0.008513  0.000019   
2      2  0.308958  0.112160  0.172870  0.110771  0.189936  0.000084   
3      3  0.355601  0.176941  0.226207  0.161770  0.023319  0.000034   
4      4  0.169829  0.050656  0.463231  0.191694  0.020030  0.000040   

          6         7  
0  0.000027  0.094210  
1  0.000009  0.048978  
2  0.000035  0.105187  
3  0.000017  0.056111  
4  0.000025  0.104494

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [17]:
submit.mean()

index    260824.500000
0             0.519728
1             0.074557
2             0.183136
3             0.052903
4             0.070066
5             0.047687
6             0.006879
7             0.045044
dtype: float64

In [18]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [19]:
'signate submit --competition-id=275 ./{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=275 ./submit/ball_18.csv --note feat=217_cv=1.1004234715781385'

### GCS

In [20]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/ball_18.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 18 (217) 
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1102
- Best CV score: 1.1004234715781385
- Wall time: 13min 42s
- 評価結果  : 1.3954167

#### 17 (191) 
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1100
- Best CV score: 1.1034107856339535
- Wall time: 26min 44s
- 評価結果  : 1.3963057

#### 16 (181) ... 選手IDごとの2017年のコース実績
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1178
- Best CV score: 1.1021930539837317
- Wall time: 12min 35s
- 評価結果  : 1.3993836

#### 15 (153) ... 左右で球種/コースを別に集計
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1189
- Best CV score: 1.1013869435441306
- Wall time: 11min 9s
- 評価結果  : 未提出

#### 14 (153) ... 左右で球種を別に集計
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1240
- Best CV score: 1.1021952294615331
- Wall time: 11min 34s
- 評価結果  : 1.3981900

#### 13 (157) ... 前の投球からの時間差を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1416
- Best CV score: 1.1062463941695753
- Wall time: 13min 10s
- 評価結果  : 1.4012184

#### 12 (156) ... コースの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1353
- Best CV score: 1.106942086205103
- Wall time: 27min 41s
- 評価結果  : 1.4016077

#### 11 (143) ... ボールカウントの集計を追加
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1209
- Best CV score: 1.1075153721728936
- Wall time: 23min 19s
- 評価結果  : 未提出

#### 10_tune (135) ... 再チューニング
- 'num_leaves' : 15,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1260
- Best CV score: 1.1118897437428692
- Wall time: 8min 40s
- 評価結果  : 1.4060397

#### 10s (135) ... trainを75%サンプリング
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1412
- Best CV score: 1.1190287120710383
- Wall time: 12min 5s
- 評価結果  : 1.4310869

#### 10 (135) ... 5相当に戻した
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1112
- Best CV score: 1.111265000190272
- Wall time: 9min 30s
- 評価結果  : 1.4008712

#### 9 (136) ... 5相当に戻した=>NG
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1106
- Best CV score: 1.1113177029272225
- Wall time: 25min 44s (local)
- 評価結果  : 1.5133521

#### 8 (157)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1100
- Best CV score: 1.1093728644236824
- Wall time: 10min 4s
- 評価結果  : 1.4925832

#### 7 (235)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1204
- Best CV score: 1.108590082437519
- Wall time: 12min 56s
- 評価結果  : 1.4946937

#### 6 (207)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1135
- Best CV score: 1.1086837935965734
- Wall time: 11min 54s
- 評価結果  : 1.3999194

#### 5-tune (参考)
- 'num_leaves' : 18,
- 'learning_rate' : 0.05,
- Best num_boost_round: 2395
- Best CV score: 1.1103399384119979
- Wall time: 18min 29s (n1-standard-8)

#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780
- CV Wall time: 27min 48s (local) -> 18min 24s (n1-standard-4) -> 9min 3s (n1-standard-8)

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806